### PR TITLE
170 specify parameters declared on start

### LINKED ISSUE #170: create syntax for variables to be specified at bot start
As an RPA-developer I would like to specify which parameters should be declared on bot start.

---

### :thought_balloon: &nbsp; Hints & Thoughts
[Check this out](https://robocorp.com/docs/setup/robot-structure#specifying-environment-variables-when-running-the-robot-in-robocorp-lab-and-the-robocorp-vs-code-extension) for parsing

### :question: &nbsp; Additional Questions to Answer / Consider

---

## :white_check_mark: &nbsp; Conditions of satisfaction

- [ ] created Syntax for specifying the parameter at runtime and document it in the wiki

### PR DESCRIPTION
**Pull request by WolfgangDaniel**
closes #170 

## Changes
For each parameter the developer has to be able to don't fill in the value for that parameter but let the user put in the value before she/he starts the robot. This is now solved. For each parameter there is a checkbox below. When this checkbox is checked than this field requires user input. So two things happen: First, the field gets disabled and the value gets erased from the field and second, the property 'requireUserInput' gets set to true for this parameter in the session storage.

## Screenshots
User Input required checked:
grafik

User Input required unchecked:
grafik

## Checklist before merge

**Developer's responsibilities**
* [ ] **Assign** one or two developers
* [ ] **Change code** if reviewer(s) has/have requested it
* [ ] **Pull request build** has passed
* [ ] **tested locally** (in at least chrome & firefox if frontend)
* [ ] updated the **documentation**
* [ ] added **tests** where necessary

### CODE REVIEW

File: client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanel.jsx [hunk 1]
```diff
@@ -23,12 +23,13 @@ const PropertiesPanel = ({
   selectedActivity,
   taskSelectionUpdated,
   disableTaskSelection,
+  robotId,
```
Comment (kej-jay) [inline client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanel.jsx:26]: linter warning

File: client/src/utils/attributeAndParamUtils.js [hunk 2]
```diff
@@ -198,22 +255,92 @@ const getParameterObject = (robotId, activityId) => {
...
+  });
+  sessionStorage.setItem(
+    'parameterLocalStorage',
+    JSON.stringify(newLocalParameterStorage)
+  );
+};
 
-    sessionStorage.setItem('parameterLocalStorage', JSON.stringify(localParametersWithoutMatch));
+/**
+ * @description Will retrieve the local parameter storage and return the current value of the userInpitRequired property
```
Comment (kej-jay) [inline client/src/utils/attributeAndParamUtils.js:323]:
```suggestion
 * @description Will retrieve the local parameter storage and return the current value of the userInputRequired property
```

File: client/src/utils/attributeAndParamUtils.js [hunk 3]
```diff
@@ -22,7 +22,7 @@ const getRobotId = () => JSON.parse(sessionStorage.getItem(ROBOT_ID_PATH));
  * @param {String} robotId the robotId ot the currently opened robot
```
Comment (kej-jay) [inline client/src/utils/attributeAndParamUtils.js:22]: I couldn't quite check if everything is right because most of is again is only prettier reformatting. I looked at the methods you added and they seemed okay.

File: client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanelComponents/PPParameterInput.jsx [hunk 4]
```diff
@@ -1,43 +1,108 @@
...
-                <Tooltip title='This field is required for the task to work'>
-                    <InfoCircleOutlined style={{ color: corporateDesign.colorBackgroundCta }} />
-                </Tooltip>
-            }
-        />
+      <Input
+        placeholder={variableName}
+        defaultValue={value}
+        value={userInputRequired ? '' : parameterValue}
+        onChange={(event) => changeParamterValue(event, variableName)}
```
Comment (kej-jay) [inline client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanelComponents/PPParameterInput.jsx:74]: Why is this needed? Previously the input needed to be confirmed through enter. Either one therefore should go. 
I have no idea, but how could possibly the performance be influenced by calling the method on every single change? Is it for certain, that the state is updated right this way or could race conditions happen?

File: client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanelComponents/PPParameterInput.jsx [hunk 5]
```diff
@@ -1,43 +1,108 @@
...
-    dataType,
-    value
-}) => (
+  onValueChange,
+  variableName,
+  isRequired,
+  // eslint-disable-next-line no-unused-vars
+  dataType,
+  value,
+  robotId,
```
Comment (kej-jay) [inline client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanelComponents/PPParameterInput.jsx:24]: As far as I can tell the robotId is passed through from all components simply to be passed in to a method by this component. Please consider loading the robotId from sessionStorage in the method itself, as you are already dealing with sessionStorage items in those methods anyway and it would help us keep the code more clean.

**Review by kej-jay — CHANGES_REQUESTED** [review]

File: client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanel.jsx (hunk 1 above)
Comment (WolfgangDaniel) [reply client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanel.jsx:26]: ✅

Comment (lukashueller) [reply client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanel.jsx:26]: FYI: The linter warning says that `robotId` is missing in props validation, so please add
```jsx
  robotId: PropTypes.string.isRequired,
```
in line 67

**Review by lukashueller — APPROVED** [review]
Looks like a really clean enhancement for me! Please fix Eriks comments and then you can merge from my point of view.

File: client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanelComponents/PPParameterInput.jsx (hunk 5 above)
Comment (WolfgangDaniel) [reply client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanelComponents/PPParameterInput.jsx:24]: Good point. I think it is not really a big problem passing the robotId as a parameter and it has the advantage that the robotId is not dependent on the session storage which is good if we might change the architecture in the future. If you are okay with it I would leave it like that for now and if the passing of properties in general gets too much we can start a refactoring issue for that :)

File: client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanelComponents/PPParameterInput.jsx [hunk 6]
```diff
@@ -1,43 +1,108 @@
...
+      variableName,
+      'requireUserInput'
+    )
+  );
+  const [parameterValue, setParameterValue] = useState(
+    parameterPropertyStatus(robotId, selectedActivity, variableName, 'value')
+  );
+
+  const changeUserInputRequirement = (event, parameterName) => {
+    if (event.target.checked) {
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanelComponents/PPParameterInput.jsx:40]: ✅ Yes of course. Thanks for the hint.

File: client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanelComponents/PPParameterInput.jsx (hunk 4 above)
Comment (WolfgangDaniel) [reply client/src/components/content/ModelerSidebar/PropertiesPanel/PropertiesPanelComponents/PPParameterInput.jsx:74]: I always thought that needing to press enter to confirm an input was just a temporary solution and here it was a convenient time for me to just fix that. The performance shouldn't be a problem as we are not working with tha backend. I haven't experienced any kind of racing conditions yet and using state like this to update a value in an input field seems to be (according to my google search and the udemy react course) a common solution :)

**Review by kej-jay — APPROVED** [review]
